xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -1566,3 +1567,448 @@ func (s) TestRingHash_ReattemptWhenGoingFromTransientFailureToIdle(t *testing.T)
 		t.Errorf("conn.GetState(): got %v, want %v", got, want)
 	}
 }
+
+// Tests that when all backends are down and then up, we may pick a TF backend
+// and we will then jump to ready backend.
+func (s) TestRingHash_TransientFailureSkipToAvailableReady(t *testing.T) {
+	emptyCallF := func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+		return &testpb.Empty{}, nil
+	}
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("Failed to create listener: %v", err)
+	}
+	restartableListener1 := testutils.NewRestartableListener(lis)
+	restartableServer1 := stubserver.StartTestService(t, &stubserver.StubServer{
+		Listener:   restartableListener1,
+		EmptyCallF: emptyCallF,
+	})
+	defer restartableServer1.Stop()
+
+	lis, err = testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("Failed to create listener: %v", err)
+	}
+	restartableListener2 := testutils.NewRestartableListener(lis)
+	restartableServer2 := stubserver.StartTestService(t, &stubserver.StubServer{
+		Listener:   restartableListener2,
+		EmptyCallF: emptyCallF,
+	})
+	defer restartableServer2.Stop()
+
+	nonExistantBackends := makeNonExistentBackends(t, 2)
+
+	const clusterName = "cluster"
+	backends := []string{restartableServer1.Address, restartableServer2.Address}
+	backends = append(backends, nonExistantBackends...)
+	endpoints := endpointResource(t, clusterName, backends)
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+	route := headerHashRoute("new_route", virtualHostName, clusterName, "address_hash")
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+	opts := []grpc.DialOption{
+		grpc.WithConnectParams(grpc.ConnectParams{
+			// Disable backoff to speed up the test.
+			MinConnectTimeout: 100 * time.Millisecond,
+		}),
+		grpc.WithResolvers(xdsResolver),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+	conn, err := grpc.NewClient("xds:///test.server", opts...)
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	if got, want := conn.GetState(), connectivity.Idle; got != want {
+		t.Errorf("conn.GetState(): got %v, want %v", got, want)
```

Comment:
Should this be a `t.Fatalf` instead?

Reply:
I changed to `t.Fatalf`. Agree it doesn't really make sense to let the test proceed in those cases.